myhoard/restore_coordinator.py

```diff
@@ -430,6 +436,66 @@ def restore_basebackup(self) -> None:
 
+    def rebuild_tables(self) -> None:
+        self._ensure_mysql_server_is_started(with_binlog=False, with_gtids=False)
+
+        def escape_identifier(name: str) -> str:
+            return "`" + name.replace("`", "``") + "`"
+
+        with self._mysql_cursor() as cursor:
+            cursor.execute(
+                "SELECT TABLE_SCHEMA,TABLE_NAME,TABLE_ROWS,AVG_ROW_LENGTH"
+                " FROM INFORMATION_SCHEMA.TABLES WHERE ENGINE='InnoDB'"
+            )
+            tables = cursor.fetchall()
+            tables = sorted(
+                [
+                    table
+                    for table in tables
+                    if (table["TABLE_SCHEMA"], table["TABLE_NAME"])
+                    not in {
+                        ("mysql", "innodb_index_stats"),
+                        ("mysql", "innodb_table_stats"),
+                    }
+                ],
+                key=lambda table: (table["TABLE_SCHEMA"], table["TABLE_NAME"]),
+            )
+            estimated_total_bytes = sum(table["TABLE_ROWS"] * table["AVG_ROW_LENGTH"] for table in tables)
+            estimated_progress_bytes = 0
+            self.log.info("Will rebuild %s tables, estimated total size: %s bytes", len(tables), estimated_total_bytes)
+            for table_index, table in enumerate(tables, start=1):
+                escaped_table_schema = escape_identifier(table["TABLE_SCHEMA"])
+                escaped_table_name = escape_identifier(table["TABLE_NAME"])
+                estimated_size_bytes = table["TABLE_ROWS"] * table["AVG_ROW_LENGTH"]
+                self.log.info(
+                    "Rebuilding table %s of %s:  %s.%s, estimated size: %s bytes, estimated total progress: %.1f%%",
+                    table_index,
+                    len(tables),
+                    escaped_table_schema,
+                    escaped_table_name,
+                    estimated_size_bytes,
+                    estimated_progress_bytes / max(estimated_total_bytes, 1) * 100,
+                )
+                try:
+                    cursor.execute(f"ALTER TABLE {escaped_table_schema}.{escaped_table_name} FORCE")
+                except pymysql.err.OperationalError as e:
+                    # ERROR 1148: The used command is not allowed with this MySQL version
+                    if e.args[0] == 1148:
+                        # This happens for some tables which are marked as using InnoDB but are actually not
+                        # normal tables (like mysql.innodb_index_stats and mysql.innodb_table_stats).
+                        # The known ones are already skipped but if we encounter more, we just skip them.
+                        self.log.error(
+                            "Could not rebuild %s.%s, error: %s, skipping it",
+                            escaped_table_schema,
+                            escaped_table_name,
+                            str(e),
+                        )
+                    else:
+                        raise
```

Comment:
If `myhoard` gets restarted I think it would be nice to skip errors which will happen on the table which is running ALTER TABLE already. (needs to be checked what happens in this case and if we need to actually wait for this to finish).

Also, this I am not sure myself, but maybe we should not hard stop in case of any problem with rebuilding, we don't know yet other cases and if we roll this out to many services, we might not be able to genbump due to that.
Maybe we should try to repair and instead of failing collect a metric and have alerts for those. What do you think?

Reply:
The main expected cause of failure would be a corrupted table, I'd rather have the genbump fail hard in that case, while there is still a previous node running.

If it doesn't, then we restore a possibly corrupted database, and then remove the previous node.
Nobody will look at the metrics until it's too late and corruption becomes visible.

For the restart, I'll check how MySQL behaves.

Reply:
I tried concurrent ALTER and MySQL does not raise errors, it just does both.